---
title: "Basic bulk T Cell Repertoire statistics"
format:
  html: 
    theme: flatly
    toc: true
    toc_depth: 3
    code-fold: true
    embed-resources: true
    number-sections: true
    grid:
      body-width: 1000px
      margin-width: 300px

jupyter: python3
---

This notebook was generated via the `bulk-tcrseq` pipeline developed at Johns Hopkins for use by BTC scientists. 

# Report Setup

## Load Packages

```{python}
from IPython.display import Image
import os
import datetime
import sys
import pandas as pd
import math
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import LinearSegmentedColormap
```

## Pipeline Parameters

Default inputs are overwritten at the command line in `modules/local/plot_sample.nf`
```{python, echo=false}
#| tags: [parameters]
#| echo: false

workflow_cmd='<command used to run the pipeline>'
project_name='path/to/project_name'
project_dir='path/to/project_dir'
sample_table='path/to/sample_table.csv'
sample_stats_csv='path/to/sample_stats.csv'
v_family_csv='path/to/v_family.csv'
```

```{python}
print('Project Name:          ' + project_name)
print('Workflow command:      ' + workflow_cmd)
print('Pipeline Directory:    ' + project_dir)
print('Date and time:         ' + str(datetime.datetime.now()))
```

## Importing custom plotting functions

Plotting functions are defined in `bin/plot_utils.py`.

```{python}
## custom plotting and timing functions
sys.path.append(project_dir + '/bin/')
from plot_utils import TicTocGenerator, tic, toc, plot_timecourse2, plt_combined

## create an instance of the TicTocGen generator
TicToc = TicTocGenerator()
```

## Loading data

```{python}
tic()

## reading combined repertoire statistics
df = pd.read_csv(sample_stats_csv, sep=',', header=0, 
                 names=['sample_id', 'patient_id', 'timepoint', 'origin', 'num_clones', 
                        'num_TCRs', 'simpson_index', 'simpson_index_corrected', 'clonality',
                        'num_in', 'num_out', 'num_stop', 'pct_prod', 'pct_out', 'pct_stop', 'pct_nonprod',
                        'cdr3_avg_len'])
print('-- Imported sample_stats_csv as `df`...')

## reading sample metadata
meta = pd.read_csv(sample_table, sep=',', header=0, index_col=None,
                   names=['sample_id', 'file_path', 'patient_id', 'timepoint', 'origin'])
print('-- Imported sample_table as `meta`...')

## reading V gene family usage 
v_family = pd.read_csv(v_family_csv, sep=',', header=0, index_col=None,
                       names=['patient_id', 'timepoint', 'origin', 'TCRBV01', 
                              'TCRBV02', 'TCRBV03', 'TCRBV04', 'TCRBV05', 'TCRBV06',
                              'TCRBV07', 'TCRBV08', 'TCRBV09', 'TCRBV10', 'TCRBV11',
                              'TCRBV12', 'TCRBV13', 'TCRBV14', 'TCRBV15', 'TCRBV16',
                              'TCRBV17', 'TCRBV18', 'TCRBV19', 'TCRBV20', 'TCRBV21',
                              'TCRBV22', 'TCRBV23', 'TCRBV24', 'TCRBV25', 'TCRBV26',
                              'TCRBV27', 'TCRBV28', 'TCRBV29', 'TCRBV30'])
v_family = v_family.sort_values(by=['patient_id', 'timepoint'])
print('-- Imported v_family_csv as `v_family`...')

toc()
```

# Sample level statistics v2

Below are plots showing basic T cell repertoire statistics. Each plot has a basic description about the statistic shown. However, biological interpretation of each plot is left to the user.

## Number of clones

```{python}
plt_combined(df, x_col='timepoint', y_col='num_clones', patient_col='patient_id')
```

**Figure 1. Number of clones across timepoints.** A clone is defined as a T cell with a unique CDR3 amino acid sequence. The number of clones is shown on the y-axis and 'origin_timepoint' is shown on the x-axis. Left: Boxplot, Right: matched paitent line plot.

## Clonality

```{python}
plt_combined(df, x_col='timepoint', y_col='clonality', patient_col='patient_id')
```

**Figure 2. The clonality of samples across timepoints.** Clonality is a measure of T cell clonal expansion and reflects the degree to which the sample is dominated by 1 or more T cell clones. Clonality is calculated as: *Clonality = 1 - entropy(clone_counts) / log2(number of clones)*. Left: Boxplot, Right: matched paitent line plot.

## Simpson Index

```{python}
plt_combined(df, x_col='timepoint', y_col='simpson_index_corrected', patient_col='patient_id')
```

**Figure 3. Corrected Simpson Index.** The Simpson Index is a measure of diversity that takes into account the number of clones and the relative abundance of each clone in a sample. The corrected Simpson Index is calculated as: *Corrected Simpson Index = sum(clone_counts\*(clone_counts-1)) / (number of TCRs\*(number of TCRs-1))*. Left: Boxplot, Right: matched paitent line plot.

## Percent of productive TCRs

```{python}
plt_combined(df, x_col='timepoint', y_col='pct_prod', patient_col='patient_id')
```

**Figure 4. Percent of productive TCRs.** A productive TCR is a DNA/RNA sequence that can be translated into a protein sequence, i.e. it does not contain a premature stop codon or an out of frame rearrangement. The percent of productive TCRs is calculated as: *Percent productive TCRs = number of productive TCRs / number of TCRs*. Left: Boxplot, Right: matched paitent line plot.

## Average CDR3 Length

```{python}
plt_combined(df, x_col='timepoint', y_col='cdr3_avg_len', patient_col='patient_id')
```

**Figure 5. Average CDR3 Length** The average length of the CDR3 region of the TCR. The CDR3 region is the most variable region of the TCR and is the region that determines antigen specificity. Left: Boxplot, Right: matched paitent line plot.

## V gene family usage

```{python}
# Loop through each patient and plot gene usage
# Calculate the number of rows and columns
N = len(v_family['patient_id'].unique())
rows = math.ceil(N / 2)
cols = 2

# Create a figure and axes for subplots
patients = v_family['patient_id'].unique()
fig, axes = plt.subplots(rows, cols, figsize=(10, rows*5), sharey=True)

for i, ax in enumerate(axes.flatten()):
    if i >= len(patients):
        ax.set_visible(False)
        continue
    
    pt = v_family[v_family['patient_id'] == patients[i]]
    pt.loc[:, ['identity']] = pt['origin'] + '_' + pt['timepoint']
    
    # removing metadata columns and set new index
    pt_raw = pt.drop(['patient_id', 'timepoint', 'origin'], axis=1)
    pt_raw.set_index('identity', inplace=True)

    # calculate percentages for each gene from the row total
    pt_pct = pt_raw.apply(lambda row: row / row.sum(), axis=1)

    #### ================ PLOTTING ================ ####

    # Create a color palette
    colors = sns.cubehelix_palette(n_colors=len(pt_pct.columns),
                               start = 1.75, rot = 1, reverse=True)
    cmap1 = LinearSegmentedColormap.from_list("my_colormap", colors)

    # Plot stacked bar chart for current patient
    pt_pct.plot(ax=ax, kind='bar', stacked=True, colormap=cmap1, 
                edgecolor='white')
    ax.set_title(str(patients[i]))
    ax.set_ylabel('Portion of TCRs')
    ax.get_legend().set_visible(False)

    # Align x-axis plots
    ax.set_xticks(range(len(pt_pct.index)))
    ax.set_xticklabels(pt_pct.index, fontsize=12,
                       rotation=0, ha='center')
    ax.set_xlabel('')  # Remove x-label 'identity'

# Add a common title for the whole figure
fig.suptitle('V Family Usage for All Patients', fontsize=24)

# Create a common legend outside the loop
handles, labels = ax.get_legend_handles_labels()
fig.legend(handles, labels, loc='center left', bbox_to_anchor=(1.0, 0.5))

# Adjust spacing between subplots
plt.tight_layout()
fig.subplots_adjust(top=0.96)

plt.savefig('v_family_usage.png')
plt.show()
plt.close()
```

**Figure 6. V gene family usage.** The V gene family usage of the TCRs in each sample. The x-axis shows the timepoint collected for each individual, and the y-axis shows the proportion of TCRs that use each V gene family.

# Execution environment

```{python}
print(os.system('mamba env export --no-builds'))
```
